### PR TITLE
oracled2: Allow user to configure external executor endpoint

### LINKED ISSUE #1687: Allow user to configure external executor endpoint

### PR DESCRIPTION
**Pull request by perimeko**
Fixed #1687

### CODE REVIEW

File: chain/cmd/bandoracled2/executor.go [hunk 1]
```diff
@@ -10,15 +12,16 @@ type executor interface {
...
+type lambdaExecutor struct {
+	URL string
+}
 
 func (e *lambdaExecutor) Execute(
 	l *Logger, exec []byte, timeout time.Duration, arg string,
 ) ([]byte, uint32) {
-	// TODO: Make URL configurable
 	result, err := byteexec.RunOnAWSLambda(
 		exec, timeout, arg,
```
Comment (sorawit) [inline chain/cmd/bandoracled2/executor.go]: nit: Let's pack all these into online now that the arguments are quite short.

File: chain/cmd/bandoracled2/run.go [hunk 2]
```diff
@@ -82,6 +82,14 @@ func runCmd(c *Context) *cobra.Command {
 				return err
 			}
 			l := NewLogger(allowLevel)
+			executorName, executorURL, err := ParseExecutor(cfg.Executor)
+			if err != nil {
+				return err
+			}
+			c.executor, err = NewExecutor(executorName, executorURL)
```
Comment (sorawit) [inline chain/cmd/bandoracled2/run.go]: This assumes that all executor strings always have two parts: the name and the URL, which is likely not always the case. I think we can just have `NewExecutor` function just takes `cfg.Executor` directly.

File: chain/cmd/bandoracled2/run.go [hunk 3]
```diff
@@ -93,12 +101,14 @@ func runCmd(c *Context) *cobra.Command {
 	cmd.Flags().String(flags.FlagChainID, "", "chain ID of BandChain network")
 	cmd.Flags().String(flags.FlagNode, "tcp://localhost:26657", "RPC url to BandChain node")
 	cmd.Flags().String(flagValidator, "", "validator address")
+	cmd.Flags().String(flagExecutor, "lambda:https://dmptasv4j8.execute-api.ap-southeast-1.amazonaws.com/bash-execute", "executor")
```
Comment (sorawit) [inline chain/cmd/bandoracled2/run.go]: Please have a more descriptive description.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/cmd/bandoracled2/executor.go [hunk 4]
```diff
@@ -10,20 +12,41 @@ type executor interface {
...
 	if err != nil {
 		l.Error(":skull: LambdaExecutor failed with error: %s", err.Error())
 		return []byte("EXECUTION_ERROR"), 255
 	}
 
 	return result, 0
 }
+
+// NewExecutor returns executor by name and executer URL
+func NewExecutor(executer string) (executor, error) {
```
Comment (sorawit) [inline chain/cmd/bandoracled2/executor.go]: typo: executor.

File: chain/cmd/bandoracled2/executor.go [hunk 5]
```diff
@@ -10,20 +12,41 @@ type executor interface {
...
+	switch name {
+	case "lambda":
+		return &lambdaExecutor{URL: url}, nil
+	default:
+		return nil, fmt.Errorf("Invalid executor name: %s, url: %s", name, url)
+	}
+}
+
+// ParseExecutor returns parsed executor string
+func ParseExecutor(executorStr string) (name string, url string, err error) {
```
Comment (sorawit) [inline chain/cmd/bandoracled2/executor.go]: This function can now be private (lowercase) now that it's not intended to be used outside.

File: chain/cmd/bandoracled2/executor.go [hunk 6]
```diff
@@ -10,20 +12,41 @@ type executor interface {
...
+	}
+	switch name {
+	case "lambda":
+		return &lambdaExecutor{URL: url}, nil
+	default:
+		return nil, fmt.Errorf("Invalid executor name: %s, url: %s", name, url)
+	}
+}
+
+// ParseExecutor returns parsed executor string
```
Comment (sorawit) [inline chain/cmd/bandoracled2/executor.go]: parseExecutor splits the executor string in the form of "name:url" into parts.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/cmd/bandoracled2/main.go [hunk 7]
```diff
@@ -26,6 +27,7 @@ type Config struct {
 	Validator string `mapstructure:"validator"`  // The validator address that I'm responsible for
 	GasPrices string `mapstructure:"gas-prices"` // Gas prices of the transaction
 	LogLevel  string `mapstructure:"log-level"`  // Log level of the logger
+	Executor  string `mapstructure:"executor"`   // Executer name and Executor URL (example: "Executor name:URL")
```
Comment (sorawit) [inline chain/cmd/bandoracled2/main.go]: typo

File: chain/cmd/bandoracled2/executor.go [hunk 8]
```diff
@@ -10,20 +12,41 @@ type executor interface {
...
+	result, err := byteexec.RunOnAWSLambda(exec, timeout, arg, e.URL)
 	if err != nil {
 		l.Error(":skull: LambdaExecutor failed with error: %s", err.Error())
 		return []byte("EXECUTION_ERROR"), 255
 	}
 
 	return result, 0
 }
+
+// NewExecutor returns executor by name and executer URL
```
Comment (sorawit) [inline chain/cmd/bandoracled2/executor.go]: typo.

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**